crates/wasi-common/src/hostcalls_impl/fs.rs

```diff
@@ -288,18 +288,18 @@ pub(crate) unsafe fn fd_fdstat_get(
 }
 
 pub(crate) unsafe fn fd_fdstat_set_flags(
-    wasi_ctx: &WasiCtx,
+    wasi_ctx: &mut WasiCtx,
     fd: wasi::__wasi_fd_t,
     fdflags: wasi::__wasi_fdflags_t,
 ) -> Result<()> {
     trace!("fd_fdstat_set_flags(fd={:?}, fdflags={:#x?})", fd, fdflags);
 
     let fd = wasi_ctx
-        .get_fd_entry(fd)?
-        .as_descriptor(0, 0)?
-        .as_os_handle();
+        .get_fd_entry_mut(fd)?
+        .as_descriptor_mut(0, 0)?
+        .as_file_mut()?;
```

Comment:
`EBADF` will be returned for stdout/stdin, as per https://github.com/bytecodealliance/wasmtime/blob/9fdc356f66c0b327bf6dda36d9cd69fd7dd93f95/crates/wasi-common/src/fdentry.rs#L32-L37

On the other hand, `fcntl` on stdin/stdout seems to be used in the wild, for instance [here](https://gist.github.com/tatsuya6502/72599e2d9785ec7c7d5e50959d2a2333)

Reply:
@marmistrz I'm not sure this is relevant for this PR as, prove me wrong here, @peterhuene PR didn't really touch the functionality of the host-common parts except for returning a mutable handle to the underlying FD. What I'm trying to say here is that the problem you're pointing at there will be present in our Unix implementation as well and perhaps it'd be better to handle it in a separate PR?

Reply:
Ah, wait, there is a change from `as_os_handle` to `as_file_mut`. Is this change intentional here? Anyway, apart from this, both versions still don't handle the stdout/in/err cases which I propose we handle in a subsequent PR.

Reply:
It was intentional, but I wasn't aware that it would exclude stdin/stdout/stderr.

I can revert but it looks like we'll need some other mechanism for getting at a mutable handle.

Reply:
I reckon it's fine to leave your change in as-is for now, as it doesn't really impact much at this very moment. We should consider more cases such as stdin/out/err or streams/sockets in a subsequent PR(s) though!

Reply:
What are your thoughts in this approach?

Changing back to `as_os_handle` to allow `fd_fdstat_set_flags` to continue to work off raw descriptors so that it doesn't impact the Unix implementation at all (i.e. it will still work with any file descriptor).

Instead, have the host impl return `Result<Option<OsHandle>>` instead of `Result<()>` to indicate the call was successful and here's the handle you need to replace the current descriptor with.  The Unix implementation would return `None`, and thus nothing changes, while the Windows one will return a new descriptor and replace what's currently stored.

Reply:
E.g.

```rust
pub(crate) unsafe fn fd_fdstat_set_flags(
    wasi_ctx: &mut WasiCtx,
    _memory: &mut [u8],
    fd: wasi::__wasi_fd_t,
    fdflags: wasi::__wasi_fdflags_t,
) -> Result<()> {
    trace!("fd_fdstat_set_flags(fd={:?}, fdflags={:#x?})", fd, fdflags);

    let descriptor = wasi_ctx
        .get_fd_entry_mut(fd)?
        .as_descriptor_mut(wasi::__WASI_RIGHTS_FD_FDSTAT_SET_FLAGS, 0)?;

    if let Some(new_handle) =
        hostcalls_impl::fd_fdstat_set_flags(&descriptor.as_os_handle(), fdflags)?
    {
        *descriptor = Descriptor::OsHandle(new_handle);
    }

    Ok(())
}
```

Reply:
This is the implementation I pushed up.